File: oss/clawpipe/sdk/src/cache.ts
/**
 * ReasoningBank Cache — client-side in-memory prompt cache.
 *
 * Hash-based deduplication with TTL expiry. Identical prompts
 * return cached results in microseconds instead of calling the LLM.
 */

interface CacheEntry {
  value: string;
  createdAt: number;
  hits: number;
}

interface CacheStats {
  size: number;
  hits: number;
  misses: number;
  hitRate: string;
  totalSaved: number;
}

export class Cache {
  private store = new Map<string, CacheEntry>();
  private ttlMs: number;
  private maxEntries: number;
  private totalHits = 0;
  private totalMisses = 0;

  constructor(ttlMs = 300_000, maxEntries = 10_000) {
    this.ttlMs = ttlMs;
    this.maxEntries = maxEntries;
  }

  /** Generate a cache key from prompt and options. */
  key(prompt: string, options: object = {}): string {
    const raw = JSON.stringify({ prompt, options });
    return this.hash(raw);
  }

  /** Get a cached value. Returns null if not found or expired.
   * Pass `{forceRefresh: true}` to bypass cache and invalidate the entry. */
  get(cacheKey: string, opts: { forceRefresh?: boolean } = {}): string | null {
    if (opts.forceRefresh) {
      this.store.delete(cacheKey);
      this.totalMisses++;
      return null;
    }
    const entry = this.store.get(cacheKey);
    if (!entry) {
      this.totalMisses++;
      return null;
    }

    if (Date.now() - entry.createdAt > this.ttlMs) {
      this.store.delete(cacheKey);
      this.totalMisses++;
      return null;
    }

    entry.hits++;
    this.totalHits++;
    return entry.value;
  }

  /** Store a value in cache. */
  set(cacheKey: string, value: string): void {
    this.evictIfFull();
    this.store.set(cacheKey, {
      value,
      createdAt: Date.now(),
      hits: 0,
    });
  }

  /** Check if a key exists and is not expired. */
  has(cacheKey: string): boolean {
    const entry = this.store.get(cacheKey);
    if (!entry) return false;
    if (Date.now() - entry.createdAt > this.ttlMs) {
      this.store.delete(cacheKey);
      return false;
    }
    return true;
  }

  /** Remove a specific entry. */
  delete(cacheKey: string): boolean {
    return this.store.delete(cacheKey);
  }

  /** Clear all cached entries. */
  clear(): void {
    this.store.clear();
    this.totalHits = 0;
    this.totalMisses = 0;
  }

  /** Get cache performance stats. */
  stats(): CacheStats {
    const total = this.totalHits + this.totalMisses;
    const hitRate = total > 0 ? ((this.totalHits / total) * 100).toFixed(1) : '0.0';
    return {
      size: this.store.size,
      hits: this.totalHits,
      misses: this.totalMisses,
      hitRate: `${hitRate}%`,
      totalSaved: this.totalHits,
    };
  }

  /** Remove expired entries. */
  prune(): number {
    const now = Date.now();
    let removed = 0;
    for (const [k, entry] of this.store) {
      if (now - entry.createdAt > this.ttlMs) {
        this.store.delete(k);
        removed++;
      }
    }
    return removed;
  }

  /** Simple string hash using djb2 algorithm. */
  private hash(input: string): string {
    let hash = 5381;
    for (let i = 0; i < input.length; i++) {
      hash = ((hash << 5) + hash + input.charCodeAt(i)) & 0xffffffff;
    }
    return `cp_${(hash >>> 0).toString(36)}`;
  }

  /** Evict least-hit entries when cache is full. */
  private evictIfFull(): void {
    if (this.store.size < this.maxEntries) return;

    const entries = Array.from(this.store.entries())
      .sort((a, b) => a[1].hits - b[1].hits);

    const toRemove = Math.ceil(this.maxEntries * 0.1);
    for (let i = 0; i < toRemove && i < entries.length; i++) {
      this.store.delete(entries[i][0]);
    }
  }
}

File: oss/clawpipe/python-sdk/clawpipe/cache.py
"""ReasoningBank Cache -- client-side in-memory prompt cache.

Hash-based deduplication with TTL expiry and LRU eviction.
"""

from __future__ import annotations

import json
import math
import time
from dataclasses import dataclass, field
from typing import Any, Optional


@dataclass
class CacheEntry:
    """A single cache entry."""

    value: str
    created_at: float
    hits: int = 0


@dataclass
class CacheStats:
    """Cache performance statistics."""

    size: int
    hits: int
    misses: int
    hit_rate: str
    total_saved: int


class Cache:
    """In-memory prompt cache with TTL and LRU eviction."""

    def __init__(self, ttl_ms: int = 300_000, max_entries: int = 10_000) -> None:
        self._store: dict[str, CacheEntry] = {}
        self._ttl_ms = ttl_ms
        self._max_entries = max_entries
        self._total_hits = 0
        self._total_misses = 0

    def key(self, prompt: str, options: Any = None) -> str:
        """Generate a cache key from prompt and options."""
        raw = json.dumps({"prompt": prompt, "options": options or {}}, sort_keys=True)
        return self._hash(raw)

    def get(self, cache_key: str) -> Optional[str]:
        """Get a cached value. Returns None if missing or expired."""
        entry = self._store.get(cache_key)
        if entry is None:
            self._total_misses += 1
            return None

        elapsed_ms = (time.time() - entry.created_at) * 1000
        if elapsed_ms > self._ttl_ms:
            del self._store[cache_key]
            self._total_misses += 1
            return None

        entry.hits += 1
        self._total_hits += 1
        return entry.value

    def set(self, cache_key: str, value: str) -> None:
        """Store a value in cache."""
        self._evict_if_full()
        self._store[cache_key] = CacheEntry(
            value=value,
            created_at=time.time(),
            hits=0,
        )

    def has(self, cache_key: str) -> bool:
        """Check if a key exists and is not expired."""
        entry = self._store.get(cache_key)
        if entry is None:
            return False
        elapsed_ms = (time.time() - entry.created_at) * 1000
        if elapsed_ms > self._ttl_ms:
            del self._store[cache_key]
            return False
        return True

    def delete(self, cache_key: str) -> bool:
        """Remove a specific entry."""
        if cache_key in self._store:
            del self._store[cache_key]
            return True
        return False

    def clear(self) -> None:
        """Clear all cached entries."""
        self._store.clear()
        self._total_hits = 0
        self._total_misses = 0

    def stats(self) -> CacheStats:
        """Get cache performance stats."""
        total = self._total_hits + self._total_misses
        hit_rate = f"{(self._total_hits / total * 100):.1f}" if total > 0 else "0.0"
        return CacheStats(
            size=len(self._store),
            hits=self._total_hits,
            misses=self._total_misses,
            hit_rate=f"{hit_rate}%",
            total_saved=self._total_hits,
        )

    def prune(self) -> int:
        """Remove expired entries."""
        now = time.time()
        expired = [
            k
            for k, entry in self._store.items()
            if (now - entry.created_at) * 1000 > self._ttl_ms
        ]
        for k in expired:
            del self._store[k]
        return len(expired)

    @staticmethod
    def _hash(input_str: str) -> str:
        """djb2 hash algorithm."""
        h = 5381
        for c in input_str:
            h = ((h << 5) + h + ord(c)) & 0xFFFFFFFF
        return f"cp_{h:x}"

    def _evict_if_full(self) -> None:
        """Evict least-hit entries when cache is full."""
        if len(self._store) < self._max_entries:
            return
        entries = sorted(self._store.items(), key=lambda x: x[1].hits)
        to_remove = math.ceil(self._max_entries * 0.1)
        for i in range(min(to_remove, len(entries))):
            del self._store[entries[i][0]]
